Update delegatesFocus to match recent spec changes

rniwa opened the pull request:
#### c1607d1c6ed9aa5cda088ed74349c3018dba4b11
<pre>
Update delegatesFocus to match recent spec changes
<a href="https://bugs.webkit.org/show_bug.cgi?id=232564">https://bugs.webkit.org/show_bug.cgi?id=232564</a>

Reviewed by Darin Adler.

Implemented the new behavior of focusDelegates and added the support for autofocus focus delegation.
Also fixed a bug that an element with autofocus content attribute inside a shadow tree gets focus.

See <a href="https://github.com/whatwg/html/pull/7079">https://github.com/whatwg/html/pull/7079</a>

* LayoutTests/fast/shadow-dom/autofocus-in-shadow-tree-autofocus-host-expected.txt: Added.
* LayoutTests/fast/shadow-dom/autofocus-in-shadow-tree-autofocus-host.html: Added.
* LayoutTests/fast/shadow-dom/autofocus-in-shadow-tree-expected.txt: Added.
* LayoutTests/fast/shadow-dom/autofocus-in-shadow-tree.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/focus-visible-020.html: Merge the test fix
posted in <a href="https://github.com/web-platform-tests/wpt/pull/35257.">https://github.com/web-platform-tests/wpt/pull/35257.</a>
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/focus/focus-autofocus-expected.txt: Rebaselined.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/focus/focus-method-delegatesFocus-expected.txt: Ditto.

* Source/WebCore/dom/Element.cpp:
(WebCore::shouldAutofocus): Fixed the bug that this wasn&apos;t ignoring elements inside shadow trees.
Autofocus attribute yields autofocus candidates if the relevant element is **inserted** into a document,
not connected to it. See <a href="https://html.spec.whatwg.org/multipage/interaction.html#the-autofocus-attribute">https://html.spec.whatwg.org/multipage/interaction.html#the-autofocus-attribute</a>

(WebCore::autoFocusDelegate): Added. This implements autofocus delegation mechanism for shadow root/host.

(WebCore::focusDelegateFromShadowHost): Renamed from findFirstProgramaticallyFocusableElementInComposedTree
and made it recursively traverse shadow trees instead of the composed/flat tree.

(WebCore::Element::focus):

Canonical link: <a href="https://commits.webkit.org/252901@main">https://commits.webkit.org/252901@main</a>
</pre>

darinadler commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3209,13 +3210,34 @@ static bool isProgramaticallyFocusable(Element& element)
+        if (auto root = shadowRootWithDelegatesFocus(element)) {
+            auto target = autoFocusDelegate(*root);
+            if (target)
+                return target;
```
Scope the target in the if statement?
```
    if (auto target = autoFocusDelegate(*root))
        return target;
```

darinadler commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3209,13 +3210,34 @@ static bool isProgramaticallyFocusable(Element& element)
+        if (auto root = shadowRootWithDelegatesFocus(element)) {
+            auto target = focusDelegateFromShadowHost(*root);
+            if (target)
+                return target;
```
Ditto.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -205,8 +205,9 @@ static bool shouldAutofocus(const Element& element)
 
     auto& document = element.document();
-    if (!element.isConnected() || !document.hasBrowsingContext())
+    if (!element.isInShadowTree() || !element.isConnected() || !document.hasBrowsingContext())
```
Clearly, this should read `element.isInShadowTree()` instead! That's why tests are failing.

darinadler replied:
Thought about it more, and we should write `if (!element.isInDocumentTree() || !document.hasBrowsingContext())`

rniwa replied:
Yeah, I've considered that but perhaps people not familiar with shadow DOM wouldn't know being inside a document tree is different from being connected to a document.

darinadler replied:
Sure, but writing it out as `isInShadowTree() || !element.isConnected()` doesn’t make that any better in my opinion. I personally think it’s easier to explain "you need to use `isInDocumentTree` instead of `isConnected` here" is actually easier to explain than "also check if it’s in a shadow tree".

rniwa replied:
Okay, that's a fair point. Will update.

rniwa committed:
Update delegatesFocus to match recent spec changes
https://bugs.webkit.org/show_bug.cgi?id=232564

Reviewed by Darin Adler.

Implemented the new behavior of focusDelegates and added the support for autofocus focus delegation.
Also fixed a bug that an element with autofocus content attribute inside a shadow tree gets focus.

See https://github.com/whatwg/html/pull/7079

* LayoutTests/fast/shadow-dom/autofocus-in-shadow-tree-autofocus-host-expected.txt: Added.
* LayoutTests/fast/shadow-dom/autofocus-in-shadow-tree-autofocus-host.html: Added.
* LayoutTests/fast/shadow-dom/autofocus-in-shadow-tree-expected.txt: Added.
* LayoutTests/fast/shadow-dom/autofocus-in-shadow-tree.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/focus-visible-020.html: Merge the test fix
posted in https://github.com/web-platform-tests/wpt/pull/35257.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/focus/focus-autofocus-expected.txt: Rebaselined.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/focus/focus-method-delegatesFocus-expected.txt: Ditto.

* Source/WebCore/dom/Element.cpp:
(WebCore::shouldAutofocus): Fixed the bug that this wasn't ignoring elements inside shadow trees.
Autofocus attribute yields autofocus candidates if the relevant element is **inserted** into a document,
not connected to it. See https://html.spec.whatwg.org/multipage/interaction.html#the-autofocus-attribute

(WebCore::autoFocusDelegate): Added. This implements autofocus delegation mechanism for shadow root/host.

(WebCore::focusDelegateFromShadowHost): Renamed from findFirstProgramaticallyFocusableElementInComposedTree
and made it recursively traverse shadow trees instead of the composed/flat tree.

(WebCore::Element::focus):

Canonical link: https://commits.webkit.org/252901@main
